src/ui/profile/ProfileSelection.js

```diff
@@ -0,0 +1,170 @@
+// eslint-disable-next-line no-redeclare
+/* global chrome */
+import React, { useEffect, useState } from 'react';
+import PropTypes from 'prop-types';
+import styles from '../styles/profile.module.css';
+import AddProfile from './AddProfile';
+import BeyondLoader from './BeyondLoader';
+import Profiles from './Profile';
+
+const LoadedProfile = (profileModel) => ({
+  status: 'loaded',
+  profile: profileModel,
+});
+
+const ErrorProfile = (error, profile) => ({
+  status: 'error',
+  error,
+  profile,
+});
+
+const LoadingProfile = (profile) => ({
+  status: 'loading',
+  profile,
+});
+
+function ProfileSelection(props) {
+  const { onCharacterReady } = props;
+  const [profiles, setProfiles] = useState(new Map());
+  const [currentProfile, setCurrentProfile] = useState(null);
+  const [errorMessage, setErrorMessage] = useState('');
+
+  useEffect(() => {
+    chrome.storage.local.get('savedProfiles', (result) => {
+      if (!result.savedProfiles) return;
+
+      const keyPairs = result.savedProfiles.map((profile) => [profile.id, LoadedProfile(profile)]);
+      setProfiles(new Map(keyPairs));
+    });
+  }, []);
+
+  const onAddProfile = (id, profilePromise) => {
+    if (profiles.has(id)) {
```

Comment:
Updates to handle promises correctly and to handle `profiles` as an object instead of a map:

```js
const onAddProfile = async (id, profilePromise) => {
    if (profiles[id]) {
        throw new Error('ALREADY_EXISTS');
    }

    // Add placeholder profile with id until loaded
    setProfiles(obj => ({ ...obj, [id]: LoadingProfile({ id }) })); // <-- using object spread syntax makes this way more succinct

    try {
        const profile = await profilePromise;
        
        // using new useChromeLocalStorage()
        setProfiles({ ...profiles, [id]: LoadedProfile(profile) });

        // using old way
        setProfiles(obj => {
            const newObj = { ...obj, [id]: LoadedProfile(profile) };
            chrome.storage.local.set({ savedProfiles: newObj });
            return newObj;
        });
    } catch (err) {
        // Remove the unloadable profile
        setProfiles(obj => {
            const { [id]: unloadable, ...newObj } = obj; // <-- object rest syntax, newObj has all the keys from obj except id
            return newObj;
        });
        setErrorMessage(err.message);
        throw err;
    }
};
```

You can see how much more succinct and easier this is to read with these changes.